### PR TITLE
storeapi: resume snap downloads

### PR DESCRIPTION
**Pull request by sergiusens**
Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

### CODE REVIEW

**Comment by sergiusens** [comment]
ok, seems we are doing the right thing
```
Starting new HTTPS connection (1): public.apps.ubuntu.com
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Starting new HTTPS connection (1): 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-24T21:12:06Z&h=d17cbd8892be4542d2a8e7bdbffa079d73001bd4 HTTP/1.1" 200 83349504
Retries left to download 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': 5.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-24T21:12:08Z&h=72ada5b9c36d65fdef3b7beed1bef4465188d1a4 HTTP/1.1" 200 83349504
Retries left to download 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': 4.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-24T21:12:09Z&h=be65420324c9ee8c88bdd14fa1c87afcb06478c5 HTTP/1.1" 200 83349504
Retries left to download 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': 3.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-24T21:12:11Z&h=833102254c063dd1502a1f0f6a2fec634bed6659 HTTP/1.1" 200 83349504
Retries left to download 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': 2.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-24T21:12:13Z&h=6418ec910dbbd9bcd7bd002c8687a069b141614b HTTP/1.1" 200 83349504
Retries left to download 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': 1.
("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

**Comment by sergiusens** [comment]
I will edit the logged debug message so we get the thrown exception in there too

**Comment by sergiusens** [comment]
```
Starting new HTTPS connection (1): search.apps.ubuntu.com
"GET /api/v1/snaps/details/core?channel=stable&fields=status%2Canon_download_url%2Cdownload_url%2Cdownload_sha3_384%2Cdownload_sha512%2Csnap_id%2Crevision%2Crelease HTTP/1.1" 200 573
Downloading core
Starting new HTTPS connection (1): public.apps.ubuntu.com
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Starting new HTTPS connection (1): 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:16Z&h=c1cda6c141defcbf04819467b55117e2a6863cff HTTP/1.1" 200 83349504
Redirections for 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': https://068ed04f23.site.internapcdn.net/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:16Z&h=c1cda6c141defcbf04819467b55117e2a6863cff
Error while downloading: ChunkedEncodingError(ProtocolError("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer')),). Retries left to download: 5.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:24Z&h=bf88e0517fb1390ed5671c7b0ebf0612b18a9d3e HTTP/1.1" 200 83349504
Redirections for 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': https://068ed04f23.site.internapcdn.net/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:24Z&h=bf88e0517fb1390ed5671c7b0ebf0612b18a9d3e
Error while downloading: ChunkedEncodingError(ProtocolError("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer')),). Retries left to download: 4.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:26Z&h=cbd0753f009f7d9ca8a7a4f18b1fd6032d2f04ae HTTP/1.1" 200 83349504
Redirections for 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': https://068ed04f23.site.internapcdn.net/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:26Z&h=cbd0753f009f7d9ca8a7a4f18b1fd6032d2f04ae
Error while downloading: ChunkedEncodingError(ProtocolError("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer')),). Retries left to download: 3.
"GET /anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap HTTP/1.1" 302 0
Resetting dropped connection: 068ed04f23.site.internapcdn.net
"GET /download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:28Z&h=81eb231649cba42085e9a5ab0ded3b756f0c9881 HTTP/1.1" 200 83349504
Redirections for 'https://public.apps.ubuntu.com/anon/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap': https://068ed04f23.site.internapcdn.net/download-snap/99T7MUlRhtI3U0QFgl5mXXESAiSwt776_1689.snap?t=2017-05-25T22:36:28Z&h=81eb231649cba42085e9a5ab0ded3b756f0c9881
```

File: snapcraft/storeapi/__init__.py [hunk 1]
```diff
@@ -292,7 +292,11 @@ def _download_snap(self, name, channel, arch, download_path,
         not_downloaded = True
         retry_count = 5
         while not_downloaded and retry_count:
-            request = self.cpi.get(download_url, stream=True)
+            headers = {}
+            if os.path.exists(download_path):
+                headers['Range'] = 'StartPos: %d'.format(
+                    os.path.getsize(download_path))
```
Comment (cprov) [inline snapcraft/storeapi/__init__.py]: Internap has `Accept-Ranges: bytes`, so we should be using `Range: bytes=<size>-'`.

See the full workflow in action with curl in https://pastebin.canonical.com/189327/

File: snapcraft/internal/indicators.py [hunk 2]
```diff
@@ -49,19 +49,29 @@ def _init_progress_bar(total_length, destination, message=None):
...
+def download_requests_stream(request_stream, destination, message=None,
+                             total_read=0):
     """This is a facility to download a request with nice progress bars."""
 
     # Doing len(request_stream.content) may defeat the purpose of a
     # progress bar
     total_length = 0
     if not request_stream.headers.get('Content-Encoding', ''):
         total_length = int(request_stream.headers.get('Content-Length', '0'))
+        # Content-Lenght in the case of resuming will be
```
Comment (kyrofa) [inline snapcraft/internal/indicators.py]: s/Lenght/Length/

File: snapcraft/internal/indicators.py [hunk 3]
```diff
@@ -49,19 +49,29 @@ def _init_progress_bar(total_length, destination, message=None):
...
+        # Content-Length - total_read so we add back up to have the feel of
+        # resuming
+        if os.path.exists(destination):
+            total_length += total_read
```
Comment (kyrofa) [inline snapcraft/internal/indicators.py]: Nice, this is to make the progress bar start partially completed, essentially?

File: snapcraft/storeapi/__init__.py [hunk 4]
```diff
@@ -287,17 +287,36 @@ def _download_snap(self, name, channel, arch, download_path,
             return
         logger.info('Downloading {}'.format(name, download_path))
 
+        resume_possible = False
+        total_read = 0
+        probe_url = requests.head(download_url)
+        if probe_url.is_redirect:
+            download_url = probe_url.headers['Location']
+            resume_possible = True
```
Comment (kyrofa) [inline snapcraft/storeapi/__init__.py]: Can you explain this? Why is a resume possible only if the download URL is a redirect?

File: snapcraft/storeapi/__init__.py [hunk 5]
```diff
@@ -287,17 +287,36 @@ def _download_snap(self, name, channel, arch, download_path,
...
         # HttpAdapter cannot help here as this is a stream.
         # LP: #1617765
         not_downloaded = True
         retry_count = 5
         while not_downloaded and retry_count:
-            request = self.cpi.get(download_url, stream=True)
+            headers = {}
+            if resume_possible and os.path.exists(download_path):
+                total_read = os.path.getsize(download_path)
+                headers['Range'] = 'bytes={}-'.format(total_read)
```
Comment (kyrofa) [inline snapcraft/storeapi/__init__.py]: The equals sign doesn't appear to be part of the format (at least according to [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range)). To what spec are you referring?

File: snapcraft/internal/indicators.py (hunk 3 above)
Comment (sergiusens) [inline snapcraft/internal/indicators.py]: yes

File: snapcraft/storeapi/__init__.py (hunk 4 above)
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: it is the only place I know where `Accept-Ranges: bytes` works.

File: snapcraft/storeapi/__init__.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: look at @cprov's comment above.

Comment (cprov) [inline snapcraft/storeapi/__init__.py]: Right, the specification does not require it, space should be fine, perhaps it's a specialty of internap :-/ 

Anyway, as I've noted in the previous comment, this is how it's currently working, but it's worth noting that it contains internap hacks. Other than that 👍

**Comment by sergiusens** [comment]
@kyrofa and @cprov one last final thing to not do the wrong thing in the future if you don't mind taking a look

**Pull request merged**